esp-hal/src/rmt.rs

```diff
@@ -250,59 +250,235 @@ pub enum Error {
     MemoryBlockNotAvailable,
 }
 
-///  Convenience trait to work with pulse codes.
-pub trait PulseCode: crate::private::Sealed {
-    /// Create a new instance
-    fn new(level1: Level, length1: u16, level2: Level, length2: u16) -> Self;
+///  Convenience newtype to work with pulse codes.
+#[derive(Clone, Copy, Default, Eq, PartialEq)]
+#[repr(transparent)]
+pub struct PulseCode(u32);
```

Comment:
It's always been strange to me that `PulseCode` contains two pulse codes.
One would expect `PulseCode` to be just a `level` and a `length`.
Then maybe have a `CompactPulseCodes` type that is a `u32`.

In practice, I'm not sure if this non-compact `PulseCode` type is actually useful. Maybe for iterator type APIs.

Reply:
I think that might be the result of many protocols using pulse codes with alternating `Level::Low` and `Level::High`, mapping single bits to a 2-field, `u32`, pulse code.

For that reason, iterator type APIs do in fact work very well with the 2-field pulse codes, shifting out one bit for each `u32` slot in the hardware buffer. See for example the [bytes encoder](https://docs.espressif.com/projects/esp-idf/en/latest/esp32c3/api-reference/peripherals/rmt.html#bytes-encoder) in IDF; as part of #3509 I've implementing something similar in Rust.

Reply:
There could of course be use cases that would better be served by `u16`, single-field pulse codes. Probably, it doesn't make too much sense to develop that API without having an actual use in mind?